### PR TITLE
modify fpu reduction depending on visited policy

### PR DESCRIPTION
**Pull request by Eddh**
This is a fpu reduction method that seems like it is as strong as parent's net_eval for weak network while being as strong as flat fpu reduction of 0.22 for strong networks.

### CODE REVIEW

**Comment by WhiteHalmos** [comment]
Thanks for making the pull request. I'm testing out ringmaster and will see if I get some results later.

**Comment by WhiteHalmos** [comment]
```
Eddh v Next (7 games)                                   
board size: 19   komi: 7.5                              
       wins              black        white       avg cpu                                                        
Eddh      5 71.43%       2 50.00%     3 100.00%    243.99                                                        
Next      2 28.57%       0  0.00%     2  50.00%    242.35                                                        
                         2 28.57%     5  71.43%         
```

By tomorrow I should have some more statistically significant results. Config below in case anyone finds a bug in my config.

```
competition_type = 'playoff'

players = {
'Eddh' : Player('leelaz-eddh -g -w 54 --noponder -p1600 -r1'),
'Next' : Player('leelaz-next -g -w 54 --noponder -p1600 -r1'),
}

board_size = 19                                         
komi = 7.5                                              

matchups = [
    Matchup('Eddh', 'Next', alternating=True,           
            scorer='players')
]
```

**Comment by zediir** [comment]
I'm running the same test with mostly the same parameters except for -t1 and 6f274ab1, 585000 games network

**Comment by zediir** [comment]
Well that didn't take long

```
eddh6f27 v next6f27 ( 19 games)
              wins        black       white
eddh6f27   19 100.00%   10 100.00%    9 100.00%
next6f27    0  0.00%    0  0.00%    0  0.00%
                       10 52.63%    9 47.37%
```

Running  0eb8b23b next

**Comment by WhiteHalmos** [comment]
Good idea, restarting with -t1 and -d, can't believe I forgot.

**Comment by WhiteHalmos** [comment]
Now running with `validation/validation -n 54 -n 54copy -b leelaz-eddh -b leelaz-next -o "-g -p 1600 --noponder -t 1 -q -d -r 1 -w" -o "-g -p 1600 --noponder -t 1 -q -d -r 1 -w" -g 1`

Again, bugs appreciated. :)

**Comment by fell111** [comment]
For early networks, should we compare this commit with base version rather than next branch? The purpose is to use the fpu formula when we start from scratch.

**Comment by zediir** [comment]
```
eddh0eb8 v next0eb8 ( 155 games)
              wins        black       white
eddh0eb8   96 61.94%   39 65.00%   57 60.00%
next0eb8   59 38.06%   21 35.00%   38 40.00%
                       60 38.71%   95 61.29%
```

**Comment by gcp** [comment]
My result for a strong network is here:
https://github.com/gcp/leela-zero/issues/696#issuecomment-363682706

**Comment by gcp** [comment]
> The purpose is to use the fpu formula when we start from scratch.

That is a good point.

**Comment by WhiteHalmos** [comment]
My results are similar to gcp's for eddh vs next. I've stopped it for now.

```
54 wins, 42 losses                                                                                               96 games played.                                                                                                 
Status: 0 LLR 0.725129 Lower Bound -2.94444 Upper Bound 2.94444
```

**Comment by WhiteHalmos** [comment]
@Eddh would you mind rebasing again so it's easier for @gcp to pull?

Lots of data came in showing that your branch is slightly better than current next for good policy nets and much better than current next for earlier nets! This change would be a net positive compared to current next.

**Comment by Eddh** [comment]
@WhiteHalmos I will do that in a few hours.

**Comment by zediir** [comment]
Running 
```
.\leelazeddh.exe  --gpu=1 -g -p 1600 --noponder -t 1 -q -d -r 1 -w eddh6f274
vs
.\leelaznext.exe  --gpu=1 --fpu_reduction 0.0 -g -p 1600 --noponder -t 1 -q -d -r 1 -w next6f274
```

**Comment by Eddh** [comment]
I rebased and resolved the conflicts.

**Comment by WhiteHalmos** [comment]
This is for the latest d6 net, commit hashes appended.
```
board size: 19   komi: 7.5  
                        wins              black         white       avg cpu                                      
leelaz-next-b572          56 61.54%       30 65.22%     26 57.78%    257.27                                      
leelaz-next_eddh-f259     35 38.46%       19 42.22%     16 34.78%    260.31                                      
                                          49 53.85%     42 46.15%
```
<details>
<summary>Ringmaster config</summary>

```
competition_type = 'playoff'
players = {}
def LZ(binary, weights):
  players[binary] = Player(
      '~/projects/lztest/{} -p1600 --noponder -r1 -g -t1 -q -d -w {}'.format(
          binary, weights))
binary1 = 'leelaz-next-b572'
binary2 = 'leelaz-next_eddh-f259'
weight1 = 'd6'
weight2 = 'd6'
LZ(binary1, weight1)
LZ(binary2, weight2)
board_size = 19
komi = 7.5
matchups = [Matchup(binary1, binary2, alternating=True)]
```
</details>

**Comment by Eddh** [comment]
Some time ago I had done test against fpu_reduction=0 with weak net 1ac2..(137000 games net) at 800 playouts and got that : 
```
leelaz11 v leelaz_sqrt_totalvp (43/50 games)
board size: 19   komi: 7.5
                      wins              black         white       avg cpu
leelaz11                20 46.51%       14 63.64%     6  28.57%    556.76
leelaz_sqrt_totalvp     23 53.49%       15 71.43%     8  36.36%    552.37
                                        29 67.44%     14 32.56%
```

so even among weak net we seem to have very different results between 585000 and 137000 games net ?

**Comment by zediir** [comment]
Let me run 1ac2 with both 800 and 1600 playouts to see what results I get.

**Comment by Eddh** [comment]
I did some games with 1600 playouts ith 1ac2 today too : 
```
26 wins, 30 losses
56 games played.
Status: 0 LLR -0.704052 Lower Bound -2.94444 Upper Bound 2.94444
```

**Comment by WhiteHalmos** [comment]
To summarize, as I understand it, our testing is currently:

Test against current next (with fpu - 0.22) against one of the latest networks (i.e. d6f3, 6665, etc).
Test against current next (with fpu - 0) against one of the oldest networks (i.e. 1ac2)

Looking for a simple solution that improves or is about even in both cases. Correct?

**Comment by Eddh** [comment]
Yes that is also my understanding

File: src/UCTNode.cpp [hunk 1]
```diff
@@ -334,8 +343,13 @@ UCTNode* UCTNode::uct_select_child(int color) {
             continue;
         }
 
-        // get_eval() will automatically set first-play-urgency
-        auto winrate = child->get_eval(color);
+        auto winrate = 0.5f;
+        if (child->get_visits() > 0) {
+            winrate = child->get_eval(color);
+        } else { // First play urgency
+            winrate = child->get_eval(color) - cfg_fpu_reduction*std::sqrt(total_visited_policy);
```
Comment (sethtroisi) [inline src/UCTNode.cpp]: This can be cached like numerator 

auto fpu_policy_reduction = cfg_fpu_reduction * std::sqrt(total_visited_policy);

**Comment by WhiteHalmos** [comment]
@Eddh What are you using for your sprt tests and does it have documentation? I've been trying to find something like that to run on the command line for gomill results.

**Comment by Eddh** [comment]
I found this python script https://chessprogramming.wikispaces.com/Ratios%20/%20Operating%20Figures-SPRT

**Comment by WhiteHalmos** [comment]
Thanks for the [link](https://chessprogramming.wikispaces.com/Match+Statistics#toc7)!

**Comment by gcp** [comment]
```
leelaz-54bf v leelaz-eddh (851/1000 games)
board size: 19   komi: 7.5
              wins              black          white        avg cpu
leelaz-54bf    457 53.70%       210 49.41%     247 57.98%    428.20
leelaz-eddh    394 46.30%       179 42.02%     215 50.59%    429.35
                                389 45.71%     462 54.29%
```

**Comment by Eddh** [comment]
@gcp leelaz-54bf is sqrt_totalvp ?
For the very weak, I am not sure, it is not significant with 1ac2, but apparently with 6f27 it is weaker than fpu_reduction=0.0 : 
```
eddh6f27 v next6f27 ( 19 games)
              wins        black       white
eddh6f27    0  0.00%    0  0.00%    0  0.00%
next6f27   19 100.00%   10 100.00%    9 100.00%
                       10 52.63%    9 47.37%
```
(from @zediir )

**Comment by zediir** [comment]
Because I first got a different result from you I ran it again with just -g1 to see if that had something to do with it (I had gotten earlier eight straight losses and then a win when running -g8 even though earlier the result was 15/20 with -g3)

with -g3
```
eddh1ac2 v next1ac2 ( 74 games)
              wins        black       white
eddh1ac2   24 32.43%   15 34.09%    9 30.00%
next1ac2   50 67.57%   29 65.91%   21 70.00%
                       44 59.46%   30 40.54%
```

with -g1 it is still going at
```
88 wins, 105 losses
193 games played.
```

**Comment by gcp** [comment]
>Somethings up with running multiple parallel matches with validation?

Can't conclude anything from the amount of games you have.

**Comment by Eddh** [comment]
What's the plan now ? it seems this is the only option left ? Thank you all for all the test games anyway. I couldn't have done all that myself.

**Comment by WhiteHalmos** [comment]
Maybe collect the test data from various comments and come to a recommendation? Not sure what conclusions can be drawn without looking at all the data.

**Comment by gcp** [comment]
I will merge this. It's good for strong networks (previous test was better, this one is about equal) and it is better for weak ones, surely than the default, but for some networks even compared to fpu=0.0.

```
leelaz-next v leelaz-eddh (366/1000 games)
board size: 19   komi: 7.5
              wins              black          white        avg cpu
leelaz-next    189 51.64%       64  34.97%     125 68.31%    115.87
leelaz-eddh    177 48.36%       58  31.69%     119 65.03%    116.51
                                122 33.33%     244 66.67%

leelaz-next-weak v leelaz-eddh-weak (366/1000 games)
board size: 19   komi: 7.5
                   wins              black          white        avg cpu
leelaz-next-weak    104 28.42%       36  19.67%     68  37.16%    390.06
leelaz-eddh-weak    262 71.58%       115 62.84%     147 80.33%    364.04
                                     151 41.26%     215 58.74%
```

**Pull request merged**